src/Hosting/Hosting/src/Internal/HostingApplication.cs

```diff
@@ -57,12 +56,153 @@ public void DisposeContext(Context context, Exception exception)
+                    {
+                        // Not false, so create the LoggingContext if it doesn't exist and set to true.
+                        LoggingContext.EventLogEnabled = true;
+                    }
+                }
+            }
+
+            public Activity Activity
+            {
+                // Get set value or return null
+                get => _loggingContext?.Activity;
+                set
+                {
+                    if (value == null)
+                    {
+                        // If null, only set to null if the _loggingContext exists as the get will default to null.
+                        if (_loggingContext != null)
+                        {
+                            _loggingContext.Activity = null;
+                        }
+                    }
+                    else
+                    {
+                        // Not null, so create the LoggingContext if it doesn't exist and set the value.
+                        LoggingContext.Activity = value;
+                    }
+                }
+            }
+
+            internal bool HasDiagnosticListener
+            {
+                // Get set value or return false
+                get => _loggingContext?.HasDiagnosticListener ?? false;
+                set
+                {
+                    if (!value)
+                    {
+                        // If false, only set to false if the _loggingContext exists as the get will default to false.
+                        if (_loggingContext != null)
+                        {
+                            _loggingContext.HasDiagnosticListener = false;
+                        }
+                    }
+                    else
+                    {
+                        // Not false, so create the LoggingContext if it doesn't exist and set to true.
+                        LoggingContext.HasDiagnosticListener = true;
+                    }
+                }
+            }
+
+            internal HostingRequestStartingLog StartLog
+            {
+                // Get set value or return null
+                get => _loggingContext?.StartLog;
+                set
+                {
+                    if (value == null)
+                    {
+                        // If null, only set to null if the _loggingContext exists as the get will default to null.
+                        if (_loggingContext != null)
+                        {
+                            _loggingContext.StartLog = null;
+                        }
+                    }
+                    else
+                    {
+                        // Not null, so create the LoggingContext if it doesn't exist and set the value.
+                        LoggingContext.StartLog = value;
+                    }
+                }
+            }
+
+            // Get or create the _loggingContext
+            private LogContext LoggingContext => _loggingContext ??= new LogContext();
```

Comment:
`??=` @pranavkm How many points for that?

Reply:
🥇

Reply:
My remaining question is how often a Context is constructed without setting any of these properties to a non-default value.

This does seem like a *ton* of work to turn six properties into one lazy-allocated field with six properties.

Also, does this need to be thread safe? I hope not because `new LogContext()` could run multiple times if LoggingContext is accessed and initialized twice in a parallel.

Reply:
> My remaining question is how often a Context is constructed without setting any of these properties to a non-default value.

If you don't want an  `Activity` (152 bytes), 4 `ExecutionContext`s (160 bytes), 2 `string`s (76+66 bytes), 4 AsyncLocalMaps (160 bytes) and 3 Scope objects (144 bytes); => `758 bytes` per request; when you aren't outputting any of it because you aren't at `Information` level; you might set `"Microsoft.AspNetCore.Hosting.Diagnostics": "None"` which will do it.

e.g. 
```
"Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.AspNetCore.Hosting.Diagnostics": "None"
    }
},
```

And if you are picking up that extra *758 bytes per request*, what's an extra object :)

Also if you have no logging as per the TE benchmarks; which also benefits from the struct being 64 bytes rather than 82 bytes when passed.

> Also, does this need to be thread safe?

No, its contained to before and after the request in hosting and not exposed to the user code in the application

Reply:
e.g. 
![image](https://user-images.githubusercontent.com/1142958/57561696-2cbded80-7385-11e9-94cf-ef66e644d282.png)


DateTimeFormat (`string`, `byte[]`) and ConcurrentDictionary (`enumerator`) are both heartbeat rather than request allocations so should be subtracted.

Reply:
I figured there must be something more to it to go through all this effort. However, I still don't understand why lazy loading the private LogContext is necessary to reduce the allocations of `Activity` for example.

I don't understand why moving Activity from Context.Activity to Context. LoggingContext.Activity saves you from any allocations except for of the LogContext itself. None of this changes whether the Activity itself is allocated. If the Activity is allocated, it is allocated *before* the setter was ever called. And it's not like there are any structs on the LogContext...

150 lines to lazy allocating a POCO for longs, bools and reference properties seems kinda crazy.

Reply:
Yeah that doesn't save any allocations. Its because its a pass by value `struct` so is more about the size when its getting copied around the stack and to the heap in the statemachine.

Before
```
Type layout for 'Context'
Size: 48 bytes. Paddings: 6 bytes (%12 of empty space)
|======================================================================|
|   0-7: HttpContext <HttpContext>k__BackingField (8 bytes)            |
|----------------------------------------------------------------------|
|  8-15: IDisposable <Scope>k__BackingField (8 bytes)                  |
|----------------------------------------------------------------------|
| 16-23: Activity <Activity>k__BackingField (8 bytes)                  |
|----------------------------------------------------------------------|
| 24-31: HostingRequestStartingLog <StartLog>k__BackingField (8 bytes) |
|----------------------------------------------------------------------|
| 32-39: Int64 <StartTimestamp>k__BackingField (8 bytes)               |
|----------------------------------------------------------------------|
|    40: Boolean <EventLogEnabled>k__BackingField (1 byte)             |
|----------------------------------------------------------------------|
|    41: Boolean <HasDiagnosticListener>k__BackingField (1 byte)       |
|----------------------------------------------------------------------|
| 42-47: padding (6 bytes)                                             |
|======================================================================|
```

After
```
Type layout for 'Context'
Size: 16 bytes. Paddings: 0 bytes (%0 of empty space)
|===========================================================|
|   0-7: LogContext _logContext (8 bytes)                   |
|-----------------------------------------------------------|
|  8-15: HttpContext <HttpContext>k__BackingField (8 bytes) |
|===========================================================|
```


Its getting copied as `TContext` parameters in this interface
```csharp
public interface IHttpApplication<TContext>
{
    TContext CreateContext(IFeatureCollection contextFeatures);
    Task ProcessRequestAsync(TContext context);
    void DisposeContext(TContext context, Exception exception);
}
```
As well as copied to the heap in Kestrel's `ProcessRequests` statemachine as `TContext`

![image](https://user-images.githubusercontent.com/1142958/57562904-47955f80-738f-11e9-812b-6803a86ec3f5.png)

Reply:
I completely forgot that the Context itself was a struct. Too bad it's public. It would have been sooo much easier to turn the Context into a class, but that's breaking.

Reply:
And if Logging is off for Hosting then the Activity and the LogContext will never be allocated.

Reply:
> I completely forgot that the Context itself was a struct. Too bad it's public. It would have been sooo much easier to turn the Context into a class, but that's breaking.

Well `LogContext` is now a class if you can work out where to hang it rather than allocating it per request?

Reply:
> And if Logging is off for Hosting then the Activity and the LogContext will never be allocated as they are guarded by the Log Enabled check

I figured, but that's unrelated to the LogContext.

> Well LogContext is now a class if you can work out where to hang it rather than allocating it per request? (when logging or diagnostics is enabled)

Pooling is an interesting idea. If Context itself was a class, could we pool that instead?

Reply:
The `HttpContext` is "pooled" via `IHttpContextFactory` (lives on connection); but the `HostingApplication.Context` is side-car'd over it and avoids allocation by living in the stack and being passed around by value.